packages/core/src/tools/ensure-json.js

```diff
@@ -0,0 +1,21 @@
+const _ = require('lodash');
+
+const isJSONEncodable = (obj) => {
+  if (obj === null || _.isBoolean(obj) || _.isNumber(obj) || _.isString(obj)) {
+    return true;
+  }
+
+  if (!_.isPlainObject(obj) && !_.isArray(obj)) {
+    return false;
+  }
+
+  for (const key in obj) {
+    if (!isJSONEncodable(obj[key])) {
+      return false;
+    }
+  }
+
+  return true;
+};
```

Comment:
As I looked closer, I think it's necessary to provide some debug information in the error message for deeply-nested objects. So instead of returning a boolean, this function should be named `ensureJsonEncodable` and throw an error when the input value is not JSON-encodable. The debug info can just be a path showing which part of the object is unacceptable.

We should also handle circular referencing.

With that, I came up with this [example code](https://gist.github.com/eliangcs/ddff58ddda957905eac2dd85e6bc6a11) with some test cases. If that looks good to you, can you copy the function and the test cases to here? Thanks!

Reply:
Refactored, thanks for the code.